#95 PGSQL Sync

PauloMelo opened the pull request:
Here are the files for the PGSQL sync functions. In a nutshell the usage is something like this:
```
    int rows, i;
    struct kore_pgsql *pgsql;
    pgsql = kore_pgsql_init_sync();

    struct kore_buf *query;
    query = kore_buf_create(120);
    kore_buf_appendf(query, "SELECT userid, encode(usersalt,'hex'), encode(userverify,'hex') FROM tbusers WHERE userlogin='%s'", userLogin);
    if (kore_pgsql_query_sync(pgsql, (char *)query->data)) {
        rows = kore_pgsql_ntuples(pgsql);
        for (i = 0; i < rows; i++) {
            user->userID = atol(kore_pgsql_getvalue(pgsql, i, 0));
            strcpy(user->login, userLogin);
            memcpy(user->salt, kore_pgsql_getvalue(pgsql, i, 1), 16);
            memcpy(user->verify, kore_pgsql_getvalue(pgsql, i, 2), 64);
        }
    }

    kore_pgsql_cleanup(pgsql);
    kore_buf_free(query);     
```

Regards,
Paulo

jorisvink commented:
Hey,

Great. Glad you got it to work.

A few issues I have however:

* I don't see the need for kore_pgsql_init_sync(), it seems to me this can be rolled together with the actual kore_pgsql_query_sync() call. Passing a pointer to some user supplied kore_pgsql data structure (which may or may not be on the stack).

* kore_malloc() returns already zeroed memory, no need to do malloc + memset. Also do not initialize variables with function return values when you declare them on the stack.

I'll test this it in a bit, but please cleanup the points I made above and amend your pull issue with their fixes.

.joris

PS: As a sidebar, the example code you posted here has some issues I just wanted to address. The use of kore_buf here seems unnecessary as this can be replaced with an snprintf() call on a buffer declared on the stack. You also use atol() and strcpy() which aren't the best functions available for doing what you want to do. Either consider using kore_strtonum() or kore_strlcpy() for these instead if you want to avoid problems later down the line. The memcpy() functions are coping fixed sizes (16, 64) without any consideration what kore_pgsql_getvalue() returns. What if it returns less then 16 or 64? Then you are running past said buffer and copying trash.

PauloMelo commented:
Hi Joris,

well one of the reasons to need a sync is a database transaction. For that you need to use the same connection you get from the database (that is the reason I've used part of the pgsql_prepare for the connection part) and run more than one SELECT, INSERT, DELETE, UPDATE cycle. So that is the reason for using a "prepare" outside of the kore_pgsql_query_sync() call. I also used the "prepare" to create the kore_pgsql data structure just to reduce code. But if you think it is a requirement for kore to let the pointer be created outside the prepare call I can change it.

Regarding the kore_malloc() and the PS comments you are right. The sample code I sent is and old code that I migrated from a gwan servlet using a mysql database. I'm migrating from gwan to kore, well just because I like to see under the hood, and gwan is stopping... So I need something similar, and I've found your product, simple, clean code as I like it. With enough power and without unneeded complexity.

The fixed sizes are on purpose because the fields are byte arrays of fixed size used for the login creation and validation. Using the atol and strcpy is just old code... sorry for not giving you a proper kore test... I just copied my user login check that was just changed the enough code to make it work, I will improve it when I have some time.

I will fix the memset of the kore_malloc and send it up the stream.
Regards,
Paulo

jorisvink commented:
> For that you need to use the same connection you get from the database.

OK that makes sense then that you would want to reuse the kore_pgsql data structure with the same underlying connection. Since the kore_pgsql_init_sync() should stay lets make it work like this:

```
int ret;
struct kore_pgsql pgsql;

ret = kore_pgsql_async_init(&pgsql);
```

Where kore_pgsql_async_init() zeroes the incoming kore_pgsql data structure before using it and sets up the initial values in it, finds a free connection and hooks that in.

Also worth noting that there is a binary error code here, which is taken from the original pgsql code in Kore I know, which I never liked. It's very difficult to differenciate between error codes here:

* Did the call fail because we hit the maximum number of concurrent connections?
* Did the call fail in pgsql_conn_create() and do we have to look at the error member of the given kore_pgsql data structure?

I propose the Kore pgsql stuff introduces a few specific error codes that can be returned from these function calls so the caller can properly figure out what the hell went wrong.

There seems to be some cruft that sneaked into your branch as well, see the .gitignore file.

jorisvink commented:
Also worth noting is that your kore_pgsql_init_sync() returns a pointer but you are also returning integer values.

PauloMelo commented:
I've corrected the kore_pgsql_init_sync and added 2 error messages either in sync and async. Do you agree with the way I've setup the error messages?

I also cleaned the .gitignore

jorisvink commented on src/pgsql.c:
```diff
@@ -275,15 +331,19 @@ pgsql_prepare(struct kore_pgsql *pgsql, struct http_request *req,
 
-		if (!pgsql_conn_create(pgsql))
+		if (!pgsql_conn_create(pgsql)) {
+			pgsql->error = kore_strdup(PGSQL_ERROR_CONNECTING);
```
pgsql_conn_create() will set its own error under pgsql->error if it fails so no need to set the error again here.

jorisvink commented on src/pgsql.c:
```diff
@@ -75,6 +77,39 @@ kore_pgsql_init(void)
+		}
+
+		if (!pgsql_conn_create(pgsql)) {
+			pgsql->error = kore_strdup(PGSQL_ERROR_CONNECTING);
```
Same as below, pgsql_conn_create() will set its own pgsql->error value if it fails. So the assignment can be removed as well here.

jorisvink commented:
Your diff looks very OK now, minus those two things I pointed out.

If you clean up the confusing between spaces and tabs you have in same places I'll gladly merge this pull request.

Afterwards I will provide an example on how to use this in the same spirit as Kore its other examples.

.joris

jorisvink commented on src/pgsql.c:
```diff
@@ -137,6 +172,20 @@ kore_pgsql_query_params(struct kore_pgsql *pgsql, struct http_request *req,
+	/* Success only for SELECTS or COMMANDS */
+        if ((PQresultStatus(pgsql->result) != PGRES_TUPLES_OK) &&
+		(PQresultStatus(pgsql->result) != PGRES_COMMAND_OK)) {
+		pgsql->state = KORE_PGSQL_STATE_ERROR;
```
If this fails, set the pgsql->error as well in the the form of:

pgsql->error = kore_strdup(PQerrorMessage(pgsql->conn->db));

So it is available for the caller.

PauloMelo commented:
Hi Joris,

I've made the changes on vi. What is the editor that you use in order to try to keep the same look and feel.

Regarding PGSQL we still have to address the multi-connection for multi-databases.

Regards,
Paulo

jorisvink commented:
> What is the editor that you use in order to try to keep the same look and feel.

vi with its default settings.

> Regarding PGSQL we still have to address the multi-connection for multi-databases.

That might be as simple as adding the connection strings with an alias such as:

```

kore_pgsql_register(const char *alias, const char *connstring);

````

And then passing the alias to the kore_pgsql_async_init() function (or similar for the async stuff) which would then either find a connection for that alias or create a new one

PauloMelo commented:
Ok. That sounds consistent.

Do you want me to do it with you supervision?

On 22/12/2015, at 22:14, Joris Vink <notifications@github.com> wrote:

What is the editor that you use in order to try to keep the same look and feel.

vi with its default settings.

Regarding PGSQL we still have to address the multi-connection for multi-databases.

That might be as simple as adding the connection strings with an alias such as:


kore_pgsql_register(const char *alias, const char *connstring);

And then passing the alias to the kore_pgsql_async_init() function (or similar for the async stuff) which would then either find a connection for that alias or create a new one

—
Reply to this email directly or view it on GitHub.

jorisvink commented:
Go for it.

PauloMelo commented:
What do you think?
I've created the kore_pgsql_register() the way you sugested.
For sync statements I've added the dbname on the kore_pgsql_init_sync(). 
For async I've created a new kore_pgsql_set_database()

jorisvink commented:
The kore_pgsql_register() stuff looks fine. Although I think you are overcomplicating things in relation to the database names and what not.

If we considering introducing a kore_pgsql_init() function for both sync/async this could take the dbname. This function would function similar to what you have in your diff as kore_pgsql_init_sync() but would initialize the kore_pgsql data structure, find an existing connection for the given database name or create a new one for said database name (pgsql_conn) and would have to be called for kore sync/async queries.

There's only two places the database names should be present:
* In the pgsql_conn data structure so we know what it is connected against
* In a list of database names and their matching connection strings (like you have in your diff today)

I suggest you do this:
* introduce a common kore_pgsql_init(struct kore_pgsql *pgsql, const char *db) function which sets up the kore_pgsql data structure, finds a connection for the given database (or creates a new one if there's still slots available). This way a bunch of duplicate code for finding a connection can be removed.

* remove kore_pgsql_init_sync() and all of the additional code you introduced for finding/setting the database stuff.

* If kore_pgsql_init() finds a suitable connectiong the pgsql_prepare() function can also be simplified as its looking for a connection right now as well.

.joris

PauloMelo commented:
Ok.

should I rename the actual kore_pgsql_init() to kore_pgsql_init_module() ?

jorisvink commented:
> should I rename the actual kore_pgsql_init() to kore_pgsql_init_module() ?

No need.

PauloMelo commented:
you suggest to convert the kore_pgsql_init_sync() to kore_pgsql_init() but there are conflicting types. The existing kore_pgsql_init is to be run once per worker and is used to initialize pools and lists.

how do you want me to proceed? set the new name for kore_pgsql_db_init()?

jorisvink commented:
Hi,

Sorry for the delay, holidays and such.

I peeked at the diff and it seems quite alright at this stage. However your comments in the actual code that you left behind need to be addressed.

I think overall the approach we are taking here for the "sync" stuff vs the async stuff that exists needs to be homogenised a bit.

I recommend the following we introduce the following init function for both sync and async queries:

```
int kore_pgsql_query_init(struct kore_pgsql *psql, struct http_request *req, const char *dbname, int async);
```

Which takes the kore_pgsql data structure passed, initialised it, finds an available db connection to use (or in case of async queues it up if none was found yet). For sync queries req should be NULL.

The async parameter should have 2 constants defined for it:
```
#define KORE_PGSQL_SYNC     0
#define KORE_PGSQL_ASYNC    1
```

After said initialisation we are primed to execute the query:

```
int kore_pgsql_query(struct kore_pgsql *pgsq, const char *query);
```

This seems more sensible from an API point of view.
I will probably merge in your stuff after new year and then rework it a bit to add these things to it.

.joris

PauloMelo commented:
Yes I agree Joris.
I've made the suggested changes.

.paulo

jorisvink commented:
I just committed your diff and several changed by myself. See https://github.com/jorisvink/kore/commit/a281fd57130c9e9cdee4eb31029734cd51ce6d00

jorisvink commented:
Committed with changes in https://github.com/jorisvink/kore/commit/a281fd57130c9e9cdee4eb31029734cd51ce6d00
